packages/artifact/src/internal/download-http-client.ts

```diff
@@ -0,0 +1,189 @@
+    console.log(rawResponse)
+    throw new Error(`Unable to get ContainersItems from ${resourceUrl}`)
+  }
+
+  /**
+   * Concurrently downloads all the files that are part of an artifact
+   * @param downloadItems information about what items to download and where to save them
+   */
+  async downloadSingleArtifact(downloadItems: DownloadItem[]): Promise<void> {
+    const DOWNLOAD_CONCURRENCY = getDownloadFileConcurrency()
+    // limit the number of files downloaded at a single time
+    const parallelDownloads = [...new Array(DOWNLOAD_CONCURRENCY).keys()]
+    this.downloadHttpManager.createClients(DOWNLOAD_CONCURRENCY)
+    let downloadedFiles = 0
+    await Promise.all(
+      parallelDownloads.map(async index => {
+        while (downloadedFiles < downloadItems.length) {
+          const currentFileToDownload = downloadItems[downloadedFiles]
+          downloadedFiles += 1
+          await this.downloadIndividualFile(
+            index,
+            currentFileToDownload.sourceLocation,
+            currentFileToDownload.targetPath
+          )
+        }
+      })
+    )
+
+    // done downloading, safety dispose all connections
+    this.downloadHttpManager.disposeAllConnections()
+  }
+
+  /**
+   * Downloads an individual file
+   * @param httpClientIndex the index of the http client that is used to make all of the calls
+   * @param artifactLocation origin location where a file will be downloaded from
+   * @param downloadPath destination location for the file being downloaded
+   */
+  private async downloadIndividualFile(
+    httpClientIndex: number,
+    artifactLocation: string,
+    downloadPath: string
+  ): Promise<void> {
+    const stream = fs.createWriteStream(downloadPath)
+    const client = this.downloadHttpManager.getClient(httpClientIndex)
+    const requestOptions = getRequestOptions('application/octet-stream', true)
+    const response = await client.get(artifactLocation, requestOptions)
+    let isGzip = false
+    if (
+      response.message.headers['content-encoding'] &&
+      response.message.headers['content-encoding'] === 'gzip'
+    ) {
+      isGzip = true
+    }
+    if (isSuccessStatusCode(response.message.statusCode)) {
+      await this.pipeResponseToStream(response, stream, isGzip)
+    } else if (isRetryableStatusCode(response.message.statusCode)) {
+      warning(
+        `Received http ${response.message.statusCode} during file download, will retry ${artifactLocation} after 10 seconds`
+      )
+      // if an error is encountered, dispose of the http connection, wait, and create a new one
+      this.downloadHttpManager.disposeClient(httpClientIndex)
+      await new Promise(resolve =>
+        setTimeout(resolve, getRetryWaitTimeInMilliseconds())
+      )
+      this.downloadHttpManager.replaceClient(httpClientIndex)
+      const retryResponse = await client.get(artifactLocation)
+      if (isSuccessStatusCode(retryResponse.message.statusCode)) {
+        if (
```

Comment:
This logic is duplicated with line 119 above.  You could have created a private method to reduce duplication. Try apply DRY principle as much as possible.  :) 
```
var isGzip = (response) => {
...
}
```

and call `pipeResponseToStream(.., .., isGzip(response))`

Reply:
Nice private method 😀